From ac7db44532f006748e86fa27879fe1c8a3945f7c Mon Sep 17 00:00:00 2001
From: Adam Duskett <adam.duskett@amarulasolutions.com>
Date: Mon, 9 Sep 2024 12:27:59 +0000
Subject: [PATCH] package/mender: bump package version to 4.0.4

A complete rewrite of the package, as it is now a cmake package.

  - Removed non-systemd support, and mender now depends on systemd.

  - Add several .service files in package/mender/files. These files are from
    the meta-mender yocto repository.

  - mender uses several submodules, and the tags from github do not include
    them. As such, the version is set to the tag commit and SUBMODULES = YES
    is set.

  - Several new options are included (See Config.in)

  - package/mender/readme.txt file is updated.

  - New dependencies are as follows:
    - boost
    - libarchive
    - libglib2
    - lmdb
    - systemd
    - util-linux

Signed-off-by: Adam Duskett <adam.duskett@amarulasolutions.com>
---
 board/mender/x86_64/linux.config              |   2 +
 board/mender/x86_64/post-build.sh             |  10 +-
 board/mender/x86_64/post-image-efi.sh         |  30 ++-
 configs/mender_x86_64_efi_defconfig           |   6 +-
 ...ptional-dynamic-linking-to-boost-log.patch |  36 ++++
 package/mender/Config.in                      | 133 ++++++++++--
 package/mender/S42mender                      |  60 ------
 package/mender/artifact_info                  |   1 -
 package/mender/device_type                    |   1 -
 package/mender/files/mender-authd.service     |  14 ++
 .../files/mender-client-data-dir.service      |  13 ++
 .../files/mender-set-systemd-machine-id.sh    |  40 ++++
 .../files/mender-systemd-growfs-data.service  |  13 ++
 .../files/mender-systemd-machine-id.service   |  12 ++
 package/mender/files/mender-updated.service   |  14 ++
 package/mender/mender-client.service          |  24 ---
 package/mender/mender.conf                    |   1 +
 package/mender/mender.hash                    |  48 ++---
 package/mender/mender.mk                      | 198 ++++++++++--------
 package/mender/readme.txt                     |   7 +-
 support/testing/tests/package/test_mender.py  |  46 ++--
 .../tests/package/test_mender/post-build.sh   |  30 +++
 .../rootfs-overlay/etc/mender/mender.conf     |  11 +
 23 files changed, 508 insertions(+), 242 deletions(-)
 create mode 100644 package/mender/0001-add-optional-dynamic-linking-to-boost-log.patch
 delete mode 100644 package/mender/S42mender
 delete mode 100644 package/mender/artifact_info
 delete mode 100644 package/mender/device_type
 create mode 100644 package/mender/files/mender-authd.service
 create mode 100644 package/mender/files/mender-client-data-dir.service
 create mode 100644 package/mender/files/mender-set-systemd-machine-id.sh
 create mode 100644 package/mender/files/mender-systemd-growfs-data.service
 create mode 100644 package/mender/files/mender-systemd-machine-id.service
 create mode 100644 package/mender/files/mender-updated.service
 delete mode 100644 package/mender/mender-client.service
 create mode 100755 support/testing/tests/package/test_mender/post-build.sh
 create mode 100644 support/testing/tests/package/test_mender/rootfs-overlay/etc/mender/mender.conf

diff --git a/board/mender/x86_64/linux.config b/board/mender/x86_64/linux.config
index f17fc18edb..8c27f2ff25 100644
--- a/board/mender/x86_64/linux.config
+++ b/board/mender/x86_64/linux.config
@@ -59,6 +59,8 @@ CONFIG_EXT4_FS=y
 CONFIG_FUSE_FS=y
 CONFIG_VFAT_FS=y
 CONFIG_SQUASHFS=y
+CONFIG_TMPFS=y
+CONFIG_TMPFS_POSIX_ACL=y
 CONFIG_NLS_CODEPAGE_437=y
 CONFIG_NLS_ISO8859_1=y
 CONFIG_UNWINDER_FRAME_POINTER=y
diff --git a/board/mender/x86_64/post-build.sh b/board/mender/x86_64/post-build.sh
index 23d1120d45..bf135364ea 100755
--- a/board/mender/x86_64/post-build.sh
+++ b/board/mender/x86_64/post-build.sh
@@ -1,7 +1,5 @@
 #!/usr/bin/env bash
 set -e
-DEVICE_TYPE="buildroot-x86_64"
-ARTIFACT_NAME="1.0"
 
 function parse_args {
     local o O opts
@@ -16,10 +14,12 @@ function parse_args {
             shift 2
             ;;
         (-d|--device-type)
-            DEVICE_TYPE="${2}"; shift 2
+             # Ignored to have same options as other scripts
+             shift 2
             ;;
         (-a|--artifact-name)
-            ARTIFACT_NAME="${2}"; shift 2
+            # Ignored to have same options as other scripts
+            shift 2
             ;;
         (--)
             shift; break
@@ -47,8 +47,6 @@ function mender_fixup {
 function main {
     parse_args "${@}"
     mender_fixup
-    echo "device_type=${DEVICE_TYPE}" > "${TARGET_DIR}/etc/mender/device_type"
-    echo "artifact_name=${ARTIFACT_NAME}" > "${TARGET_DIR}/etc/mender/artifact_info"
 }
 
 main "${@}"
diff --git a/board/mender/x86_64/post-image-efi.sh b/board/mender/x86_64/post-image-efi.sh
index 97b9e06d21..e8479f6759 100755
--- a/board/mender/x86_64/post-image-efi.sh
+++ b/board/mender/x86_64/post-image-efi.sh
@@ -31,9 +31,36 @@ function parse_args {
     done
 }
 
+# Generate a mender bootstrap artifact.
+# See https://northerntech.atlassian.net/browse/MEN-2585
+generate_mender_bootstrap_artifact() {
+  set -x
+  bootstrap_artifact="bootstrap.mender"
+  data_part="${BINARIES_DIR}"/data-part
+
+  rm -rf "${BINARIES_DIR:?}"/"${bootstrap_artifact}"
+  rm -rf "${data_part:?}"
+
+  img_checksum=$(sha256sum "${BINARIES_DIR}"/rootfs.ext4 |awk '{print $1}')
+
+  "${HOST_DIR}"/bin/mender-artifact write bootstrap-artifact \
+    --compression none \
+    --artifact-name "${ARTIFACT_NAME}" \
+    --device-type "${DEVICE_TYPE}" \
+    --provides "rootfs-image.version:${ARTIFACT_NAME}" \
+    --provides "rootfs-image.checksum:${img_checksum}" \
+    --clears-provides "rootfs-image.*" \
+    --output-path "${BINARIES_DIR}"/"${bootstrap_artifact}" \
+    --version 3
+
+    cp -rf "${BINARIES_DIR}"/"${bootstrap_artifact}" "${data_part}/"
+    echo "device_type=${DEVICE_TYPE}" > "${BINARIES_DIR}"/data-part/device_type
+}
+
 # Create the data partition
 function make_data_partition {
     "${HOST_DIR}/sbin/mkfs.ext4" \
+        -d "${BINARIES_DIR}"/data-part \
         -F \
         -r 1 \
         -N 0 \
@@ -47,7 +74,7 @@ function make_data_partition {
 function generate_mender_image {
     echo "Creating ${BINARIES_DIR}/${DEVICE_TYPE}-${ARTIFACT_NAME}.mender"
     "${HOST_DIR}/bin/mender-artifact" \
-        --compression lzma \
+        --compression none \
         write rootfs-image \
         -t "${DEVICE_TYPE}" \
         -n "${BR2_VERSION}" \
@@ -63,6 +90,7 @@ function generate_image {
 # Main function.
 function main {
     parse_args "${@}"
+    generate_mender_bootstrap_artifact
     make_data_partition
     generate_image
     generate_mender_image
diff --git a/configs/mender_x86_64_efi_defconfig b/configs/mender_x86_64_efi_defconfig
index b2b30af832..7c1064b98a 100644
--- a/configs/mender_x86_64_efi_defconfig
+++ b/configs/mender_x86_64_efi_defconfig
@@ -8,8 +8,8 @@ BR2_PACKAGE_HOST_LINUX_HEADERS_CUSTOM_6_6=y
 BR2_TARGET_GENERIC_GETTY_PORT="ttyS0"
 BR2_ROOTFS_DEVICE_CREATION_DYNAMIC_EUDEV=y
 
-# Required as vda4 doesn't mount on first boot with busybox
-BR2_INIT_SYSV=y
+# Systemd is a requirement of Mender.
++BR2_INIT_SYSTEMD=y
 
 # Required tools to create bootable media
 BR2_PACKAGE_HOST_DOSFSTOOLS=y
@@ -29,7 +29,7 @@ BR2_PACKAGE_HOST_MENDER_ARTIFACT=y
 # Filesystem / image
 BR2_TARGET_ROOTFS_EXT2=y
 BR2_TARGET_ROOTFS_EXT2_4=y
-BR2_TARGET_ROOTFS_EXT2_SIZE="128M"
+BR2_TARGET_ROOTFS_EXT2_SIZE="192M"
 # BR2_TARGET_ROOTFS_TAR is not set
 BR2_ROOTFS_OVERLAY="board/mender/x86_64/overlay"
 BR2_ROOTFS_POST_BUILD_SCRIPT="board/mender/x86_64/post-build.sh"
diff --git a/package/mender/0001-add-optional-dynamic-linking-to-boost-log.patch b/package/mender/0001-add-optional-dynamic-linking-to-boost-log.patch
new file mode 100644
index 0000000000..cba65eeb98
--- /dev/null
+++ b/package/mender/0001-add-optional-dynamic-linking-to-boost-log.patch
@@ -0,0 +1,36 @@
+From 9b2d4ca5aaacddc23032c33f56cd3d2d86aaefb7 Mon Sep 17 00:00:00 2001
+From: Adam Duskett <adam.duskett@amarulasolutions.com>
+Date: Mon, 9 Sep 2024 08:17:03 +0000
+Subject: [PATCH] add optional dynamic linking to boost::log
+
+When building boost as a dynamic library, during linking, hundreds of
+`undefined reference to `boost::log` errors occure. As the default behavior
+currently is to statically link against boost::log, it is best to add a new
+option that explicitly enables dynamically linking to boost::log. This new
+option is called MENDER_USE_DYNAMIC_BOOST_LOG.
+
+Upstream: https://github.com/mendersoftware/mender/pull/1663
+
+Signed-off-by: Adam Duskett <adam.duskett@amarulasolutions.com>
+---
+ cmake/boost.cmake | 4 ++++
+ 1 file changed, 4 insertions(+)
+
+diff --git a/cmake/boost.cmake b/cmake/boost.cmake
+index dfa02f2..aae1957 100644
+--- a/cmake/boost.cmake
++++ b/cmake/boost.cmake
+@@ -3,6 +3,10 @@
+ find_package(Boost 1.74 COMPONENTS log)
+ 
+ option(MENDER_DOWNLOAD_BOOST "Download Boost if it is not found (Default: OFF)" OFF)
++option(MENDER_USE_DYNAMIC_BOOST_LOG "Link dynamically to boost::log  (Default: OFF)" OFF)
++if(MENDER_USE_DYNAMIC_BOOST_LOG)
++  ADD_DEFINITIONS(-DBOOST_LOG_DYN_LINK)
++endif()
+ 
+ if(NOT MENDER_DOWNLOAD_BOOST AND NOT ${Boost_FOUND})
+   message(FATAL_ERROR
+-- 
+2.46.0
+
diff --git a/package/mender/Config.in b/package/mender/Config.in
index c9b9d0b757..faa7d10473 100644
--- a/package/mender/Config.in
+++ b/package/mender/Config.in
@@ -1,14 +1,27 @@
 config BR2_PACKAGE_MENDER
 	bool "mender"
-	depends on BR2_PACKAGE_HOST_GO_TARGET_ARCH_SUPPORTS
-	depends on BR2_PACKAGE_HOST_GO_TARGET_CGO_LINKING_SUPPORTS
-	depends on BR2_TOOLCHAIN_HAS_THREADS
-	select BR2_PACKAGE_HOST_GO
+	depends on BR2_INSTALL_LIBSTDCPP # boost
+	depends on BR2_TOOLCHAIN_HAS_THREADS_NPTL # boost-log
+	depends on BR2_TOOLCHAIN_SUPPORTS_ALWAYS_LOCKFREE_ATOMIC_INTS # boost-log
+	depends on !BR2_TOOLCHAIN_HAS_GCC_BUG_64735 # boost-log
+	depends on BR2_USE_MMU # libglib2
+	depends on BR2_USE_WCHAR # libglib2
+	depends on BR2_PACKAGE_SYSTEMD # runtime
+	select BR2_PACKAGE_BOOST
+	select BR2_PACKAGE_BOOST_FILESYSTEM
+	select BR2_PACKAGE_BOOST_LOG
+	select BR2_PACKAGE_DBUS if !BR2_PACKAGE_DBUS_BROKER
+	select BR2_PACKAGE_LIBARCHIVE
+	select BR2_PACKAGE_LIBGLIB2
+	select BR2_PACKAGE_LMDB
 	select BR2_PACKAGE_OPENSSL
 	select BR2_PACKAGE_OPENSSL_FORCE_LIBOPENSSL
 	select BR2_PACKAGE_LIBOPENSSL_ENGINES
 	select BR2_PACKAGE_LIBOPENSSL_ENABLE_MD4
 	select BR2_PACKAGE_LIBOPENSSL_ENABLE_RMD160
+	select BR2_PACKAGE_SYSTEMD_RESOLVED # runtime
+	select BR2_PACKAGE_UTIL_LINUX # runtime
+	select BR2_PACKAGE_UTIL_LINUX_BINARIES # flock
 	help
 	  Mender is an open source over-the-air (OTA) software updater
 	  for embedded Linux devices. Mender comprises a client
@@ -21,16 +34,13 @@ config BR2_PACKAGE_MENDER
 	    - uboot-tools must be selected.
 	    - fw_printenv must be functional on the device.
 	  The Mender Uboot patches are at:
-	  https://github.com/mendersoftware/meta-mender/
+	  https://github.com/mendersoftware/meta-mender/tree/master/meta-mender-core/recipes-bsp/u-boot/patches
 
 	  For more information on Manual U-Boot integration, see:
-	  https://docs.mender.io/2.6/system-updates-yocto-project/board-integration/bootloader-support/u-boot/manual-u-boot-integration
+	  https://docs.mender.io/3.4/system-updates-yocto-project/board-integration/bootloader-support/u-boot/manual-u-boot-integration
 
 	  For systems using Grub2 as the bootloader:
 	    - Mender depends on the mender-grubenv package.
-	    - The mender-grubenv package provides a fw_printenv script,
-	      which is not compatible with the uboot-tools fw_printenv
-	      script.
 
 	  The mender project recommends using uboot to load Grub2 as a
 	  secondary bootloader whenever possible. Using Grub2 as a
@@ -39,7 +49,104 @@ config BR2_PACKAGE_MENDER
 
 	  https://github.com/mendersoftware/mender
 
-comment "mender needs a toolchain w/ threads"
-	depends on BR2_PACKAGE_HOST_GO_TARGET_ARCH_SUPPORTS
-	depends on BR2_PACKAGE_HOST_GO_TARGET_CGO_LINKING_SUPPORTS
-	depends on !BR2_TOOLCHAIN_HAS_THREADS
+if BR2_PACKAGE_MENDER
+
+config BR2_PACKAGE_MENDER_AUTH
+	bool "authentication daemon"
+	select BR2_PACKAGE_CA_CERTIFICATES
+	help
+	  Select this if your device will connect to a Mender
+	  server. This installs the following:
+	    - mender-authd.service
+	    - mender-updated.service
+	  Otherwise, mender is expected to be ran as a standalone
+	  app.
+
+config BR2_PACKAGE_MENDER_IDENTITY_SCRIPTS
+	bool "identity scripts"
+	help
+	  Install an example script called by Mender client to
+	  collect device identity data.
+
+config BR2_PACKAGE_MENDER_INVENTORY_SCRIPTS
+	bool "inventory scripts"
+	help
+	  Install several scripts used for device inventory
+	  management. These include:
+
+	    - bootloader-integration:
+	      Tries to determine which type of bootloader
+	      integration has been used.
+
+	    - hostinfo
+	      Collects information about current host.
+
+	    - network
+	      Collects the list of network interfaces, as well as
+	      ethernet and IP addresses of each of the interfaces.
+
+	    - os
+	      Returns an "os" attribute to Mender containing the
+	      currently running OS.
+
+	    - provides
+	      Returns the current list of providees for the devices,
+	      using the Mender client's `show-provides` command.
+
+	    - rootfs-type
+	      Determines what the root filesystem type is.
+
+	    - update-modules
+	      Returns the update-modules currently installed on the
+	      system in /usr/share/mender/modules/v3
+
+config BR2_PACKAGE_MENDER_INVENTORY_NETWORK_SCRIPTS
+	bool "inventory network scripts"
+	depends on BR2_PACKAGE_BUSYBOX_SHOW_OTHERS
+	select BR2_PACKAGE_WGET # runtime
+	help
+	  Install the mender-inventory-geo script which collects
+	  information on geo localization. These include:
+	  - IP
+	  - City
+	  - Country
+	  - Time zone
+
+config BR2_PACKAGE_MENDER_MODULES
+	bool "modules"
+	help
+	  Install the following modules in
+	  usr/share/mender/modules/v3:
+	    - directory
+	    - rootfs-image
+	    - script
+	    - single-file
+
+	  Install the following if the packages are selected:
+	    - docker
+	    - rpm
+
+config BR2_PACKAGE_MENDER_HOST_MODULES_GEN
+	bool "host module generation scripts"
+	help
+	  Install scripts to generate update modules.
+	  These scripts are installed to HOST_DIR
+	  as they are not meant to be ran on the target.
+	  These include:
+	    - docker-artifact-gen
+	    - directory-artifact-gen
+	    - single-file-artifact-gen
+
+endif # BR2_PACKAGE_MENDER
+
+comment "mender needs a toolchain w/ C++, NPTL, wchar"
+	depends on BR2_USE_MMU
+	depends on BR2_TOOLCHAIN_SUPPORTS_ALWAYS_LOCKFREE_ATOMIC_INTS
+	depends on !BR2_INSTALL_LIBSTDCPP || !BR2_USE_WCHAR || \
+		!BR2_TOOLCHAIN_HAS_THREADS_NPTL
+
+comment "mender needs a toolchain not affected by GCC bug 64735"
+	depends on BR2_TOOLCHAIN_HAS_GCC_BUG_64735
+
+comment "mender needs systemd"
+	depends on !BR2_PACKAGE_SYSTEMD
diff --git a/package/mender/S42mender b/package/mender/S42mender
deleted file mode 100644
index 0f2077d089..0000000000
--- a/package/mender/S42mender
+++ /dev/null
@@ -1,60 +0,0 @@
-#!/bin/sh
-#
-# Starts mender service.
-#
-DAEMON="mender"
-DAEMON_PATH="/usr/bin/mender"
-PIDFILE="/var/run/${DAEMON}.pid"
-DAEMON_ARGS="daemon"
-
-start() {
-	# If /var/lib/mender is a symlink to /var/run/mender, and
-	#   - the filesystem is RO (i.e. we can not rm the symlink),
-	#     create the directory pointed to by the symlink.
-	#   - the filesystem is RW (i.e. we can rm the symlink),
-	#     replace the symlink with an actual directory
-	if [ -L /var/lib/mender ] && [ "$(readlink /var/lib/mender)" = "/var/run/mender" ]
-	then
-		if rm -f /var/lib/mender >/dev/null 2>&1; then
-			mkdir -p /var/lib/mender
-		else
-			echo "No persistent location to store mender data. Data will be lost"
-			echo "at reboot. Are you sure this is what you want to do?"
-			mkdir -p "$(readlink /var/lib/mender)"
-		fi
-	fi
-
-	printf "Starting mender service: "
-	umask 077
-	start-stop-daemon -bmSqp "$PIDFILE" -x ${DAEMON_PATH} -- ${DAEMON_ARGS}
-	status=$?
-	[ "${status}" = 0 ] && echo "OK" || echo "FAIL"
-}
-
-stop() {
-	printf "Stopping mender service: "
-	start-stop-daemon -Kqp "$PIDFILE"
-	status=$?
-	[ "${status}" = 0 ] && echo "OK" || echo "FAIL"
-}
-restart() {
-	stop
-	start
-}
-
-case "$1" in
-	start)
-		start
-		;;
-	stop)
-		stop
-		;;
-	restart|reload)
-		restart
-		;;
-	*)
-		echo "Usage: $0 {start|stop|reload|restart}"
-		exit 1
-esac
-
-exit $?
diff --git a/package/mender/artifact_info b/package/mender/artifact_info
deleted file mode 100644
index 1c84b088d0..0000000000
--- a/package/mender/artifact_info
+++ /dev/null
@@ -1 +0,0 @@
-artifact_name=BUILDROOT_ARTIFACT
diff --git a/package/mender/device_type b/package/mender/device_type
deleted file mode 100644
index a097ef2b3a..0000000000
--- a/package/mender/device_type
+++ /dev/null
@@ -1 +0,0 @@
-device_type=BUILDROOT_DEVICE
diff --git a/package/mender/files/mender-authd.service b/package/mender/files/mender-authd.service
new file mode 100644
index 0000000000..f789efe798
--- /dev/null
+++ b/package/mender/files/mender-authd.service
@@ -0,0 +1,14 @@
+[Unit]
+Description=Mender authentication service
+After=systemd-resolved.service mender-client-data-dir.service data.mount
+Conflicts=mender.service
+
+[Service]
+Type=idle
+User=root
+Group=root
+ExecStart=/usr/bin/mender-auth daemon
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
diff --git a/package/mender/files/mender-client-data-dir.service b/package/mender/files/mender-client-data-dir.service
new file mode 100644
index 0000000000..6420838cbe
--- /dev/null
+++ b/package/mender/files/mender-client-data-dir.service
@@ -0,0 +1,13 @@
+[Unit]
+Description=Mender persistent data dir
+After=data.mount
+Before=mender-updated.service
+
+[Service]
+Type=oneshot
+User=root
+Group=root
+ExecStart=/bin/mkdir -p -m 0700 /data/mender
+
+[Install]
+WantedBy=mender-updated.service
diff --git a/package/mender/files/mender-set-systemd-machine-id.sh b/package/mender/files/mender-set-systemd-machine-id.sh
new file mode 100644
index 0000000000..eb5c26cd29
--- /dev/null
+++ b/package/mender/files/mender-set-systemd-machine-id.sh
@@ -0,0 +1,40 @@
+#!/usr/bin/env sh
+set -eu
+# Copyright 2022 Northern.tech AS
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Write the systemd machine-id to the bootloader environment
+# so that it will persist between updates.
+#
+
+BOOTENV_PRINT=fw_printenv
+BOOTENV_SET=fw_setenv
+if [ -x /usr/bin/grub-mender-grubenv-set ]; then
+  BOOTENV_PRINT=grub-mender-grubenv-print
+  BOOTENV_SET=grub-mender-grubenv-set
+fi
+
+CURRENT_BOOTLOADER_ID=$(${BOOTENV_PRINT} mender_systemd_machine_id | tee /dev/null | cut -d= -f2)
+CURRENT_SYSTEMD_ID=$(cat /etc/machine-id)
+
+rc=0
+if [ -z "${CURRENT_BOOTLOADER_ID}" ] && [ -n "${CURRENT_SYSTEMD_ID}" ]; then
+  ${BOOTENV_SET} "mender_systemd_machine_id" "${CURRENT_SYSTEMD_ID}"
+  rc=${?}
+elif [ "${CURRENT_BOOTLOADER_ID}" != "${CURRENT_SYSTEMD_ID}" ]; then
+  echo "Error; bootloader and systemd disagree on machine-id." >&2
+  rc=1
+fi
+
+exit "${rc}"
diff --git a/package/mender/files/mender-systemd-growfs-data.service b/package/mender/files/mender-systemd-growfs-data.service
new file mode 100644
index 0000000000..6be90f09e4
--- /dev/null
+++ b/package/mender/files/mender-systemd-growfs-data.service
@@ -0,0 +1,13 @@
+[Unit]
+Description=Grow File System on /data
+DefaultDependencies=no
+BindsTo=data.mount
+Conflicts=shutdown.target
+After=data.mount
+Before=shutdown.target
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=/lib/systemd/systemd-growfs /data
+TimeoutSec=0
diff --git a/package/mender/files/mender-systemd-machine-id.service b/package/mender/files/mender-systemd-machine-id.service
new file mode 100644
index 0000000000..d366fc12db
--- /dev/null
+++ b/package/mender/files/mender-systemd-machine-id.service
@@ -0,0 +1,12 @@
+[Unit]
+Description=Mender persistent machine ID for systemd
+ConditionPathExists=/etc/machine-id
+
+[Service]
+Type=oneshot
+User=root
+Group=root
+ExecStart=/usr/bin/mender-set-systemd-machine-id.sh
+
+[Install]
+WantedBy=multi-user.target
diff --git a/package/mender/files/mender-updated.service b/package/mender/files/mender-updated.service
new file mode 100644
index 0000000000..3552c6294d
--- /dev/null
+++ b/package/mender/files/mender-updated.service
@@ -0,0 +1,14 @@
+[Unit]
+Description=Mender OTA update service
+After=systemd-resolved.service mender-client-data-dir.service data.mount
+Conflicts=mender.service
+
+[Service]
+Type=idle
+User=root
+Group=root
+ExecStart=/usr/bin/mender-update daemon
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
diff --git a/package/mender/mender-client.service b/package/mender/mender-client.service
deleted file mode 100644
index 5539a56fd6..0000000000
--- a/package/mender/mender-client.service
+++ /dev/null
@@ -1,24 +0,0 @@
-[Unit]
-Description=Mender OTA update service
-After=systemd-resolved.service
-
-[Service]
-Type=idle
-User=root
-Group=root
-ExecStartPre=/bin/sh -c '\
-if [ -L /var/lib/mender \
-     -a "$(readlink /var/lib/mender)" = "/var/run/mender" ]; then \
-    if rm -f /var/lib/mender >/dev/null 2>&1; then \
-        mkdir -p /var/lib/mender; \
-    else \
-        echo "No persistent location to store mender data. Data will be lost" \
-        echo "at reboot. Are you sure this is what you want to do?"; \
-        mkdir -p "$(readlink /var/lib/mender)"; \
-    fi; \
-fi'
-ExecStart=/usr/bin/mender daemon
-Restart=on-abort
-
-[Install]
-WantedBy=multi-user.target
diff --git a/package/mender/mender.conf b/package/mender/mender.conf
index f89118ac1b..65b2781366 100644
--- a/package/mender/mender.conf
+++ b/package/mender/mender.conf
@@ -4,6 +4,7 @@
   "RetryPollIntervalSeconds": 300,
   "RootfsPartA": "/dev/mmcblk0p2",
   "RootfsPartB": "/dev/mmcblk0p3",
+  "DeviceTypeFile": "/etc/mender/device_type",
   "ServerCertificate": "/etc/mender/server.crt",
   "ServerURL": "https://docker.mender.io",
   "TenantToken": "dummy"
diff --git a/package/mender/mender.hash b/package/mender/mender.hash
index 118cce9d02..afb4f75350 100644
--- a/package/mender/mender.hash
+++ b/package/mender/mender.hash
@@ -1,45 +1,23 @@
 # Locally computed:
-sha256  8a149b11b990beb7620b372023eec28df63cee6a1aa4132e88ecef554f080f72  mender-3.4.0-go2.tar.gz
+sha256  ba8d754d7f3248fe513a18a6575f8c32fd86eed1f2bdbd11e46426a695f0d743  mender-80f37fe47f9eac22d78a8a848b77072d45976c3e-git4.tar.gz
+sha256  418bdfce185824f861177ce6a0fdeb381bd1d9e0521038c9a636de0b60742fbb  LIC_FILES_CHKSUM.sha256
 
 # Vendor licenses
 # Generated with sed '/^[A-Za-z0-9_]/s/^/sha256  /' LIC_FILES_CHKSUM.sha256
 
 # Apache-2.0 license.
-sha256  1033348db7606a7e61b6484f293847cf8d7a35766efebb97e304d4bd5d7f3f6b  LICENSE
-sha256  1033348db7606a7e61b6484f293847cf8d7a35766efebb97e304d4bd5d7f3f6b  vendor/github.com/mendersoftware/mender-artifact/LICENSE
-sha256  73ba74dfaa520b49a401b5d21459a8523a146f3b7518a833eea5efa85130bf68  vendor/github.com/mendersoftware/openssl/LICENSE
-sha256  cfc7749b96f63bd31c3c42b5c471bf756814053e847c10f3eb003417bc523d30  vendor/github.com/minio/sha256-simd/LICENSE
-sha256  8f5d89b47d7a05a199b77b7e0f362dad391d451ebda4ef48ba11c50c071564c7  vendor/github.com/mendersoftware/progressbar/LICENSE
+sha256  d0f406b04e7901e6b4076bdf5fd20f9d7f04fc41681069fd8954413ac6295688  LICENSE
 #
-# BSD 2 Clause license.
-sha256  8d427fd87bc9579ea368fde3d49f9ca22eac857f91a9dec7e3004bdfab7dee86  vendor/github.com/pkg/errors/LICENSE
-sha256  e4646a82a976369d7ae8f6ed5c11d35dc0af18433a8ccc24c85b459ad8b95128  vendor/github.com/godbus/dbus/LICENSE
-sha256  2be1b548b0387ca8948e1bb9434e709126904d15f622cc2d0d8e7f186e4d122d  vendor/github.com/gorilla/websocket/LICENSE
+# MIT clause license
+sha256  86b998c792894ccb911a1cb7994f7a9652894e7a094c0b5e45be2f553f45cf14  vendor/json/LICENSE.MIT
+sha256  08c345dca1019879171c2c465410f7bee984bbfbdb177afd2d95bf9ff6267602  vendor/json/docs/mkdocs/docs/home/license.md
+sha256  2aefa40fa7e1c0fab18fbad5b15b02406eea1399d3cca42cc1012b65b4a56f74  vendor/tiny-process-library/LICENSE
 #
-# BSD 3 Clause license.
-sha256  0d9e582ee4bff57bf1189c9e514e6da7ce277f9cd3bc2d488b22fbb39a6d87cf  vendor/github.com/klauspost/compress/LICENSE
-sha256  2eb550be6801c1ea434feba53bf6d12e7c71c90253e0a9de4a4f46cf88b56477  vendor/github.com/pmezard/go-difflib/LICENSE
-sha256  2d36597f7117c38b006835ae7f537487207d8ec407aa9d9980794b2030cbc067  vendor/golang.org/x/sys/LICENSE
-sha256  0634b008cee55ca01f0888d2f5aba2d34e66c3f52c31a4e16a5d5d33d0c2a03e  vendor/github.com/bmatsuo/lmdb-go/LICENSE.md
-sha256  87640bc4df2ceb1559f268a3db1ba859ab780f7ba5b1b4545173d4680a3d918b  vendor/github.com/remyoudompheng/go-liblzma/LICENSE
-sha256  2d36597f7117c38b006835ae7f537487207d8ec407aa9d9980794b2030cbc067  vendor/golang.org/x/term/LICENSE
+# UNLICENSE
+sha256  7e12e5df4bae12cb21581ba157ced20e1986a0508dd10d0e8a4ab9a4cf94e85c  vendor/lmdbxx/UNLICENSE
 #
-# ISC license.
-sha256  1b93a317849ee09d3d7e4f1d20c2b78ddb230b4becb12d7c224c927b9d470251  vendor/github.com/davecgh/go-spew/LICENSE
+# CC0 1.0 Universal
+sha256  a2010f343487d3f7618affe54f789f5487602331c0a8d03f49e9a7c547cf0499  vendor/expected/COPYING
 #
-# MIT license.
-sha256  6d2de1cde19c4d2bd8bcd9aaa1d581f4cfa3db9cf71896140330eaa2f2687685  vendor/github.com/klauspost/pgzip/LICENSE
-sha256  5d966570d7a442d4e969892860a914e542c97f262c873baee8f0aa48e1f40212  vendor/github.com/klauspost/cpuid/v2/LICENSE
-sha256  51a0c9ec7f8b7634181b8d4c03e5b5d204ac21d6e72f46c313973424664b2e6b  vendor/github.com/sirupsen/logrus/LICENSE
-sha256  f8e536c1c7b695810427095dc85f5f80d44ff7c10535e8a9486cf393e2599189  vendor/github.com/stretchr/testify/LICENSE
-sha256  ffa15bdce332058a03a1d923910864fb6e58bf6df66a0e3914284725b327183e  vendor/github.com/ungerik/go-sysfs/LICENSE
-sha256  da277af11b85227490377fbcac6afccc68be560c4fff36ac05ca62de55345fd7  vendor/github.com/urfave/cli/v2/LICENSE
-sha256  b2663894033a05fd80261176cd8da1d72546e25842d5c1abcc852ca23b6b61b0  vendor/github.com/stretchr/objx/LICENSE
-sha256  d18f6323b71b0b768bb5e9616e36da390fbd39369a81807cca352de4e4e6aa0b  vendor/gopkg.in/yaml.v3/LICENSE
-sha256  08eab1118c80885fa1fa6a6dd7303f65a379fcb3733e063d20d1bbc2c76e6fa1  vendor/github.com/mattn/go-isatty/LICENSE
-#
-# OpenLDAP Public License
-sha256  310fe25c858a9515fc8c8d7d1f24a67c9496f84a91e0a0e41ea9975b1371e569  vendor/github.com/bmatsuo/lmdb-go/LICENSE.mdb.md
-
-# sha256 of all the vendor licenses combined
-sha256  93d9db9b43097b174e362ac8f398aa9a37fcdde2974d0567b62bd6d1e0d22037  LIC_FILES_CHKSUM.sha256
+# Boost Software License
+sha256  c9bff75738922193e67fa726fa225535870d2aa1059f91452c411736284ad566  vendor/optional-lite/LICENSE.txt
diff --git a/package/mender/mender.mk b/package/mender/mender.mk
index ee1398244b..81f43f6d37 100644
--- a/package/mender/mender.mk
+++ b/package/mender/mender.mk
@@ -4,112 +4,140 @@
 #
 ################################################################################
 
-MENDER_VERSION = 3.4.0
-MENDER_SITE = $(call github,mendersoftware,mender,$(MENDER_VERSION))
-MENDER_LICENSE = Apache-2.0, BSD-2-Clause, BSD-3-Clause, ISC, MIT, OLDAP-2.8
+# 4.0.4
+MENDER_VERSION = 80f37fe47f9eac22d78a8a848b77072d45976c3e
+MENDER_SITE = https://github.com/mendersoftware/mender.git
+MENDER_SITE_METHOD = git
+MENDER_GIT_SUBMODULES = YES
+MENDER_LICENSE = Apache-2.0, BSL-1.0, CC0-1.0, MIT, Unlicense
 MENDER_CPE_ID_VENDOR = northern.tech
-
-# Vendor license paths generated with:
-#    awk '{print $2}' LIC_FILES_CHKSUM.sha256 | grep vendor
 MENDER_LICENSE_FILES = \
 	LICENSE \
 	LIC_FILES_CHKSUM.sha256 \
-	vendor/github.com/mendersoftware/mender-artifact/LICENSE \
-	vendor/github.com/mendersoftware/openssl/LICENSE \
-	vendor/github.com/minio/sha256-simd/LICENSE \
-	vendor/github.com/mendersoftware/progressbar/LICENSE \
-	vendor/github.com/pkg/errors/LICENSE \
-	vendor/github.com/godbus/dbus/LICENSE \
-	vendor/github.com/gorilla/websocket/LICENSE \
-	vendor/github.com/klauspost/compress/LICENSE \
-	vendor/github.com/pmezard/go-difflib/LICENSE \
-	vendor/golang.org/x/sys/LICENSE \
-	vendor/github.com/bmatsuo/lmdb-go/LICENSE.md \
-	vendor/github.com/remyoudompheng/go-liblzma/LICENSE \
-	vendor/golang.org/x/term/LICENSE \
-	vendor/github.com/davecgh/go-spew/LICENSE \
-	vendor/github.com/klauspost/pgzip/LICENSE \
-	vendor/github.com/klauspost/cpuid/v2/LICENSE \
-	vendor/github.com/sirupsen/logrus/LICENSE \
-	vendor/github.com/stretchr/testify/LICENSE \
-	vendor/github.com/ungerik/go-sysfs/LICENSE \
-	vendor/github.com/urfave/cli/v2/LICENSE \
-	vendor/github.com/stretchr/objx/LICENSE \
-	vendor/gopkg.in/yaml.v3/LICENSE \
-	vendor/github.com/mattn/go-isatty/LICENSE \
-	vendor/github.com/bmatsuo/lmdb-go/LICENSE.mdb.md
-
-MENDER_DEPENDENCIES = host-pkgconf openssl
-
-MENDER_LDFLAGS = -X github.com/mendersoftware/mender/conf.Version=$(MENDER_VERSION)
-
-MENDER_UPDATE_MODULES_FILES = \
-	directory \
-	script \
-	single-file \
-	$(if $(BR2_PACKAGE_DOCKER_CLI),docker) \
-	$(if $(BR2_PACKAGE_RPM),rpm)
-
-define MENDER_INSTALL_CONFIG_FILES
-	$(INSTALL) -d -m 755 $(TARGET_DIR)/etc/mender/scripts
-	echo -n "3" > $(TARGET_DIR)/etc/mender/scripts/version
-
-	$(INSTALL) -D -m 0644 $(MENDER_PKGDIR)/mender.conf \
-		$(TARGET_DIR)/etc/mender/mender.conf
-	$(INSTALL) -D -m 0644 $(MENDER_PKGDIR)/server.crt \
-		$(TARGET_DIR)/etc/mender/server.crt
-
-	$(INSTALL) -D -m 0755 $(@D)/support/mender-device-identity \
-		$(TARGET_DIR)/usr/share/mender/identity/mender-device-identity
-	$(foreach f,bootloader-integration hostinfo network os rootfs-type, \
-		$(INSTALL) -D -m 0755 $(@D)/support/mender-inventory-$(f) \
-			$(TARGET_DIR)/usr/share/mender/inventory/mender-inventory-$(f)
-	)
+	vendor/expected/COPYING \
+	vendor/json/docs/mkdocs/docs/home/license.md \
+	vendor/json/LICENSE.MIT \
+	vendor/lmdbxx/UNLICENSE \
+	vendor/optional-lite/LICENSE.txt \
+	vendor/tiny-process-library/LICENSE
+
+MENDER_DEPENDENCIES += \
+	host-pkgconf \
+	boost \
+	ca-certificates \
+	libarchive \
+	libglib2 \
+	lmdb \
+	openssl \
+	systemd \
+	util-linux
+
+MENDER_INSTALL_TARGET_OPTS += \
+	install-bin \
+	install-dbus \
+	install-mender-update \
+	install-rootfs-image-module \
+	install-systemd
+
+MENDER_INSTALL_SERVICE_FILES = \
+	mender-client-data-dir.service \
+	mender-systemd-growfs-data.service \
+	mender-systemd-machine-id.service\
+	mender-set-systemd-machine-id.sh
+
+MENDER_CONF_OPTS = -DMENDER_USE_DYNAMIC_BOOST_LOG=ON
+
+ifeq ($(BR2_PACKAGE_DBUS_BROKER),y)
+MENDER_DEPENDENCIES += dbus-broker
+else
+MENDER_DEPENDENCIES += dbus
+endif
+
+ifeq ($(BR2_PACKAGE_MENDER_AUTH),y)
+MENDER_INSTALL_TARGET_OPTS += install-mender-auth
+MENDER_INSTALL_SERVICE_FILES += \
+	mender-authd.service \
+	mender-updated.service
+endif
 
-	$(INSTALL) -D -m 0755 $(MENDER_PKGDIR)/artifact_info \
-			$(TARGET_DIR)/etc/mender/artifact_info
+ifeq ($(BR2_PACKAGE_MENDER_IDENTITY_SCRIPTS),y)
+MENDER_INSTALL_TARGET_OPTS += install-identity-scripts
+endif
+
+ifeq ($(BR2_PACKAGE_MENDER_INVENTORY_SCRIPTS),y)
+MENDER_INSTALL_TARGET_OPTS += install-inventory-scripts
+endif
+
+ifeq ($(BR2_PACKAGE_MENDER_INVENTORY_NETWORK_SCRIPTS),y)
+MENDER_DEPENDENCIES += wget
+MENDER_INSTALL_TARGET_OPTS += install-inventory-network-scripts
+endif
 
-	$(INSTALL) -D -m 0755 $(MENDER_PKGDIR)/device_type \
-			$(TARGET_DIR)/etc/mender/device_type
+ifeq ($(BR2_PACKAGE_MENDER_MODULES),y)
+MENDER_INSTALL_TARGET_OPTS += install-modules
+MENDER_UPDATE_MODULES_FILES_REMOVE = \
+	deb \
+	$(if $(BR2_PACKAGE_DOCKER_CLI),,docker) \
+	$(if $(BR2_PACKAGE_RPM),,rpm)
 
-	mkdir -p $(TARGET_DIR)/var/lib
-	ln -snf /var/run/mender $(TARGET_DIR)/var/lib/mender
-	$(foreach f,$(MENDER_UPDATE_MODULES_FILES), \
-		$(INSTALL) -D -m 0755 $(@D)/support/modules/$(notdir $(f)) \
-			$(TARGET_DIR)/usr/share/mender/modules/v3/$(notdir $(f))
+define MENDER_REMOVE_UPDATE_MODULES
+	$(foreach f,$(MENDER_UPDATE_MODULES_FILES_REMOVE), \
+		$(RM) $(TARGET_DIR)/usr/share/mender/modules/v3/$(f)
 	)
 endef
+MENDER_POST_INSTALL_TARGET_HOOKS += += MENDER_REMOVE_UPDATE_MODULES
+endif
+
+ifeq ($(BR2_PACKAGE_MENDER_HOST_MODULES_GEN),y)
+define MENDER_INSTALL_MODULES_GEN
+	DESTDIR=$(HOST_DIR) \
+	$(BR2_MAKE) -C $(@D) \
+		install-modules-gen
+endef
+MENDER_POST_INSTALL_TARGET_HOOKS += += MENDER_INSTALL_MODULES_GEN
+endif
 
-MENDER_POST_INSTALL_TARGET_HOOKS += MENDER_INSTALL_CONFIG_FILES
+ifeq ($(BR2_PACKAGE_GZIP),y)
+MENDER_DEPENDENCIES += gzip
+MENDER_CONF_OPTS += -DMENDER_ARTIFACT_GZIP_COMPRESSION=ON
+else
+MENDER_CONF_OPTS += -DMENDER_ARTIFACT_GZIP_COMPRESSION=OFF
+endif
 
 ifeq ($(BR2_PACKAGE_XZ),y)
 MENDER_DEPENDENCIES += xz
+MENDER_CONF_OPTS += -DMENDER_ARTIFACT_LZMA_COMPRESSION=ON
 else
-MENDER_TAGS += nolzma
+MENDER_CONF_OPTS += -DMENDER_ARTIFACT_LZMA_COMPRESSION=OFF
 endif
 
-ifeq ($(BR2_PACKAGE_DBUS)$(BR2_PACKAGE_LIBGLIB2),yy)
-MENDER_DEPENDENCIES += libglib2
-define MENDER_INSTALL_DBUS_AUTHENTICATION_MANAGER_CONF
-	$(INSTALL) -D -m 0755 $(@D)/support/dbus/io.mender.AuthenticationManager.conf \
-		$(TARGET_DIR)/etc/dbus-1/system.d/io.mender.AuthenticationManager.conf
-
-	$(INSTALL) -D -m 0755 $(@D)/support/dbus/io.mender.UpdateManager.conf \
-		$(TARGET_DIR)/etc/dbus-1/system.d/io.mender.UpdateManager.conf
-endef
-MENDER_POST_INSTALL_TARGET_HOOKS += MENDER_INSTALL_DBUS_AUTHENTICATION_MANAGER_CONF
+ifeq ($(BR2_PACKAGE_ZSTD),y)
+MENDER_DEPENDENCIES += zstd
+MENDER_CONF_OPTS += -DMENDER_ARTIFACT_ZSTD_COMPRESSION=ON
 else
-MENDER_TAGS += nodbus
+MENDER_CONF_OPTS += -DMENDER_ARTIFACT_ZSTD_COMPRESSION=OFF
 endif
 
+define MENDER_INSTALL_MISSING_FILES
+	$(INSTALL) -D -m 0755 \
+		$(@D)/vendor/tiny-process-library/libtiny-process-library.so \
+	$(TARGET_DIR)/usr/lib/libtiny-process-library.so
+endef
+MENDER_POST_INSTALL_TARGET_HOOKS += MENDER_INSTALL_MISSING_FILES
+
 define MENDER_INSTALL_INIT_SYSTEMD
-	$(INSTALL) -D -m 0644 $(MENDER_PKGDIR)/mender-client.service \
-		$(TARGET_DIR)/usr/lib/systemd/system/mender-client.service
+	$(foreach f,$(MENDER_INSTALL_SERVICE_FILES), \
+		$(INSTALL) -D -m 0644 $(MENDER_PKGDIR)/files/$(f) \
+			$(TARGET_DIR)/usr/lib/systemd/system/$(f)
+	)
 endef
 
-define MENDER_INSTALL_INIT_SYSV
-	$(INSTALL) -D -m 755 $(MENDER_PKGDIR)/S42mender \
-		$(TARGET_DIR)/etc/init.d/S42mender
+define MENDER_INSTALL_TARGET_CMDS
+	$(TARGET_MAKE_ENV) \
+	$(TARGET_BUILD_ENV) \
+	DESTDIR=$(TARGET_DIR) \
+	$(BR2_MAKE) -C $(@D) \
+		$(MENDER_INSTALL_TARGET_OPTS)
 endef
 
-$(eval $(golang-package))
+$(eval $(cmake-package))
diff --git a/package/mender/readme.txt b/package/mender/readme.txt
index c884524a1b..8c5bac29bb 100644
--- a/package/mender/readme.txt
+++ b/package/mender/readme.txt
@@ -28,6 +28,8 @@ In Buildroot the following packages are provided:
     - This will install the client on target rootfs
 - BR2_PACKAGE_HOST_MENDER_ARTIFACT
     - This will install the 'mender-artifact' tool in host rootfs.
+- BR2_PACKAGE_MENDER_GRUBENV
+    - This package provides Grub2 integration for Mender.
 
 To fully utilize atomic image-based deployments using the A/B update
 strategy, additional integration is required in the bootloader. This
@@ -38,6 +40,9 @@ integrations please visit:
 
     https://github.com/mendersoftware/buildroot-mender
 
+There is also an example mender_x86_64_efi_defconfig that can be used
+as a demo using qemu-system-x86_64 (see board/mender/x86_64/readme.txt)
+
 Default configurations files
 ----------------------------
 
@@ -46,7 +51,7 @@ files that need your attention:
 
 - /etc/mender/mender.conf
     - main configuration file for the Mender client
-    - https://docs.mender.io/client-configuration/configuration-file/configuration-options
+    - https://docs.mender.io/client-installation/configuration-file
 
 - /etc/mender/artifact_info
     - The name of the image or update that will be built. This is what the
diff --git a/support/testing/tests/package/test_mender.py b/support/testing/tests/package/test_mender.py
index fdf55ed616..37c6647463 100644
--- a/support/testing/tests/package/test_mender.py
+++ b/support/testing/tests/package/test_mender.py
@@ -4,29 +4,51 @@ import infra.basetest
 
 
 class TestMender(infra.basetest.BRTest):
-    config = infra.basetest.BASIC_TOOLCHAIN_CONFIG + \
+    # infra.basetest.BASIC_TOOLCHAIN_CONFIG cannot be used as it does
+    # not include BR2_TOOLCHAIN_SUPPORTS_ALWAYS_LOCKFREE_ATOMIC_INTS
+    # needed by mender. This config also uses an ext4 rootfs as
+    # the default 128MB is too small for mender and systemd.
+    # a larger initrd in RAM).
+    config = \
         """
+        BR2_x86_64=y
+        BR2_x86_corei7=y
+        BR2_TOOLCHAIN_EXTERNAL=y
+        BR2_TOOLCHAIN_EXTERNAL_BOOTLIN_X86_64_CORE_I7_GLIBC_STABLE=y
+        BR2_LINUX_KERNEL=y
+        BR2_LINUX_KERNEL_CUSTOM_VERSION=y
+        BR2_LINUX_KERNEL_CUSTOM_VERSION_VALUE="6.6.50"
+        BR2_LINUX_KERNEL_USE_CUSTOM_CONFIG=y
+        BR2_LINUX_KERNEL_CUSTOM_CONFIG_FILE="board/mender/x86_64/linux.config"
         BR2_PACKAGE_MENDER=y
-        BR2_TARGET_ROOTFS_CPIO=y
+        BR2_ROOTFS_POST_BUILD_SCRIPT="{}"
         BR2_ROOTFS_OVERLAY="{}"
+        BR2_INIT_SYSTEMD=y
+        BR2_CCACHE=y
+        BR2_PACKAGE_HOST_MENDER_ARTIFACT=y
+        BR2_TARGET_ROOTFS_EXT2=y
+        BR2_TARGET_ROOTFS_EXT2_4=y
+        BR2_TARGET_ROOTFS_EXT2_SIZE="256M"
+        # BR2_TARGET_ROOTFS_TAR is not set
         """.format(
+           infra.filepath("tests/package/test_mender/post-build.sh"),
            # overlay to add a fake 'fw_printenv', used by Mender
            infra.filepath("tests/package/test_mender/rootfs-overlay"))
 
     def test_run(self):
-        cpio_file = os.path.join(self.builddir, "images", "rootfs.cpio")
-        self.emulator.boot(arch="armv5",
-                           kernel="builtin",
-                           options=["-initrd", cpio_file])
+        drive = os.path.join(self.builddir, "images", "rootfs.ext4")
+        kern = os.path.join(self.builddir, "images", "bzImage")
+        self.emulator.boot(arch="x86_64",
+                           kernel=kern,
+                           kernel_cmdline=["root=/dev/vda console=ttyS0"],
+                           options=["-cpu", "Nehalem",
+                                    "-m", "256M",
+                                    "-drive", f"file={drive},if=virtio,format=raw"])
         self.emulator.login()
 
-        # Check if the Daemon is running
-        self.assertRunOk("ls /var/run/mender.pid")
-        self.assertRunOk("ps aux | egrep [m]ender")
-
         # Check if a simple Mender command is correctly executed
-        self.assertRunOk("mender -log-level debug show-artifact")
-        self.assertRunOk("mender -log-level debug show-artifact | grep 'RUNTIME_TEST_ARTIFACT_NAME'")
+        self.assertRunOk("mender-update --log-level debug show-artifact")
+        self.assertRunOk("mender-update --log-level debug show-artifact | grep 'RUNTIME_TEST_ARTIFACT_NAME'")
         cmd = "mender show-artifact 2>&1 | grep -i 'err'"  # Check if no 'error' among the traces
         _, exit_code = self.emulator.run(cmd)
         self.assertEqual(exit_code, 1)
diff --git a/support/testing/tests/package/test_mender/post-build.sh b/support/testing/tests/package/test_mender/post-build.sh
new file mode 100755
index 0000000000..2a084e70f5
--- /dev/null
+++ b/support/testing/tests/package/test_mender/post-build.sh
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+set -e
+DEVICE_TYPE="buildroot-x86-64"
+ARTIFACT_NAME="RUNTIME_TEST_ARTIFACT_NAME"
+
+generate_mender_bootstrap_artifact() {
+  "${HOST_DIR}"/bin/mender-artifact \
+    --compression none \
+    write bootstrap-artifact \
+    --artifact-name "${ARTIFACT_NAME}" \
+    --device-type "${DEVICE_TYPE}" \
+    --provides "rootfs-image.version:${ARTIFACT_NAME}" \
+    --clears-provides "rootfs-image.*" \
+    --output-path "${TARGET_DIR}"/var/lib/mender/bootstrap.mender \
+    --version 3
+}
+
+function mender_fixup() {
+  cd "${TARGET_DIR}"
+  rm -rf var/lib/mender
+  mkdir -p var/lib/mender
+
+  rm -rf data
+  mkdir -p data
+  ln -sf /var/lib/mender/ data/
+}
+
+echo "device_type=${DEVICE_TYPE}" > "${TARGET_DIR}"/etc/mender/device_type
+mender_fixup
+generate_mender_bootstrap_artifact
diff --git a/support/testing/tests/package/test_mender/rootfs-overlay/etc/mender/mender.conf b/support/testing/tests/package/test_mender/rootfs-overlay/etc/mender/mender.conf
new file mode 100644
index 0000000000..5f423fb2cb
--- /dev/null
+++ b/support/testing/tests/package/test_mender/rootfs-overlay/etc/mender/mender.conf
@@ -0,0 +1,11 @@
+{
+  "InventoryPollIntervalSeconds": 1800,
+  "UpdatePollIntervalSeconds": 1800,
+  "RetryPollIntervalSeconds": 300,
+  "RootfsPartA": "/dev/vda2",
+  "RootfsPartB": "/dev/vda3",
+  "ServerCertificate": "/etc/mender/server.crt",
+  "ServerURL": "https://docker.mender.io",
+  "TenantToken": "dummy",
+  "DeviceTypeFile": "/etc/mender/device_type"
+}
-- 
2.46.0

